warehouse/query-core/src/main/java/datawave/query/jexl/visitors/ValidateFilterFunctionVisitor.java

```diff
@@ -0,0 +1,412 @@
+import org.apache.commons.jexl2.parser.ASTSizeFunction;
+import org.apache.commons.jexl2.parser.ASTSizeMethod;
+import org.apache.commons.jexl2.parser.ASTStringLiteral;
+import org.apache.commons.jexl2.parser.ASTTernaryNode;
+import org.apache.commons.jexl2.parser.ASTTrueNode;
+import org.apache.commons.jexl2.parser.ASTUnaryMinusNode;
+import org.apache.commons.jexl2.parser.ASTVar;
+import org.apache.commons.jexl2.parser.ASTWhileStatement;
+import org.apache.commons.jexl2.parser.JexlNode;
+import org.apache.commons.jexl2.parser.SimpleNode;
+import org.apache.log4j.Logger;
+
+import java.util.Set;
+
+import static datawave.query.jexl.functions.EvaluationPhaseFilterFunctions.EVAL_PHASE_FUNCTION_NAMESPACE;
+import static datawave.query.jexl.functions.EvaluationPhaseFilterFunctionsDescriptor.INCLUDE_TEXT;
+import static datawave.query.jexl.functions.EvaluationPhaseFilterFunctionsDescriptor.NO_EXPANSION;
+
+/**
+ * Verifies that no filter function is run against an index-only field
+ * <p>
+ * The exceptions to this rule are the #NO_EXPANSION and #TEXT functions.
+ * <p>
+ * The #NO_EXPANSION term will be pruned from the query prior to this visitor's execution, and the #TEXT query will have an index query added into the query.
+ */
+public class ValidateFilterFunctionVisitor extends BaseVisitor {
+    
+    private static final Logger log = ThreadConfigurableLogger.getLogger(ValidateFilterFunctionVisitor.class);
+    
+    private final Set<String> indexOnlyFields;
+    private final Set<String> excludedFunctions = Sets.newHashSet(NO_EXPANSION, INCLUDE_TEXT);
+    
+    public ValidateFilterFunctionVisitor(Set<String> indexOnlyFields) {
+        this.indexOnlyFields = indexOnlyFields;
+    }
+    
+    /**
+     * Generic entrypoint
+     *
+     * @param node
+     *            an arbitrary JexlNode
+     * @return the original node
+     */
+    public static JexlNode validate(JexlNode node, Set<String> indexOnlyFields) {
+        ValidateFilterFunctionVisitor visitor = new ValidateFilterFunctionVisitor(indexOnlyFields);
+        node.jjtAccept(visitor, null);
+        return node;
+    }
+    
+    @Override
+    public Object visit(ASTFunctionNode node, Object data) {
+        // visit children first, we may have function as an argument
+        node.childrenAccept(this, data);
+        validateFunction(node);
+        return data;
+    }
+    
+    private void validateFunction(ASTFunctionNode node) {
+        FunctionJexlNodeVisitor visitor = new FunctionJexlNodeVisitor();
+        node.jjtAccept(visitor, null);
+        
+        Set<String> identifiers;
+        if (visitor.namespace().equals(EVAL_PHASE_FUNCTION_NAMESPACE) && !excludedFunctions.contains(visitor.name())) {
+            for (JexlNode arg : visitor.args()) {
+                identifiers = JexlASTHelper.getIdentifierNames(arg);
+                for (String identifier : identifiers) {
+                    if (indexOnlyFields.contains(identifier)) {
```

Comment:
Possibly a silly question, but is it ever possible for indexOnlyFields to be empty? If so, then the entire visitor could simply return early without traversing the query tree since there's nothing to check against.

Reply:
After some thought I think that's a great change to make. Even though we expect index only fields to be present in nearly every case, here are few exceptions
- no index only fields generated for the date range of the query
- index only fields may be explicitly configured to be an empty set (i.e., turn off support for querying index only fields)